### ISSUE TITLE
Easier way to filter out "normal"  HTTP 4xx results from dependencies

### ISSUE THREAD

**Issue by ShaneCourtrille** [Microsoft/ApplicationInsights-dotnet-server#856]
App Insight dependency failure tracking isn't very useful for us as there are various things we use that are being logged as dependency failures but aren't.

First example was already mentioned here as #65 but was closed as UI filtering being a fix.  The problem is we can't find anyway to apply that to the Application map. 

A second example is the recording of 401 errors for a Web Reference call we make.  An HTTP 401 is being returned but that's apparently a normal part of the auth process and the call is successful.

I know we can create custom filters but that seems like a lot of work for what should be a common scenario so I'd like to see if there is an easier way to define these filters via the portal UI so it will apply to the entire App Insights ecosystem?

**Comment by MS-TimothyMothra** [comment]
I would not recommend filtering out all 401 responses because that could be a legitimate response you might want to be aware of.

I would recommend using a custom TelemetryInitializer.
https://docs.microsoft.com/en-us/azure/application-insights/app-insights-api-filtering-sampling#filtering-itelemetryprocessor
You would filter by `DependencyTelemetry`, and check for both the response code and a blacklist of URLs or Domains. Then set the `Success` property to true. This will let AppMap auto filter this item and you can keep the record of the Dependency.

**Comment by ShaneCourtrille** [comment]
I agree I don't want to filter out ALL 401 responses but from this specific domain I do because that's just how it does back/forth authentication.  My question is why should I have to write a custom TelemetryInitializer for that?  I'd expect to either be able to setup a simple filter via the portal UI (which gets sent down to the source so no data is sent) or worse case be able to setup a filter via the configuration.  

The Azure Blob scenario is even more funny to me because that is a Microsoft dependency which is designed to use 403 so I would expect that to just be handled out of the box.

**Comment by MS-TimothyMothra** [comment]
We've tried to be as generic as possible to support any services. 
A [Http 4xx status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_Client_errors) are literally an error so that's what we mark them as by default.
And we want you to have the freedom to make changes to that by writing custom TelemetryInitializers.

I'll mark this as an enhancement and our team will review a more user friendly configuration to override whitelisted domains.

**Comment by OsvaldoRosado** [comment]
This seems related to https://github.com/Microsoft/ApplicationInsights-Java/pull/606

For some time Java SDK considered 401 to be success rather than error. It's been recently removed for consistency with the rest of our SDKs. That being said, I can understand the logic behind treating 401 special in the class of 4xx errors. 

I find it difficult to justify that we call the system correctly behaving as a failure. While this applies to marking the entire 4xx range as failure, 401 is more egregiously mismarked IMO because of it's use in users who haven't authenticated yet as part of initiating the challenge in the WWW-Authenticate flow[1]. Perhaps we should reconsider this in general?

The only fair "failure" here (again IMO) is when a 401 occurs for requests where the client has attempted to provide authentication credentials [2].

[1] https://tools.ietf.org/html/rfc7235#page-6
> The 401 (Unauthorized) status code indicates that the request has not
   been applied because it lacks valid authentication credentials for
   the target resource.  The server generating a 401 response MUST send
   a WWW-Authenticate header field (Section 4.1) containing at least one
   challenge applicable to the target resource.


[2] https://tools.ietf.org/html/rfc7235#page-6
> If the request included authentication credentials, then the 401
   response indicates that authorization has been refused for those
   credentials.

**Comment by OsvaldoRosado** [comment]
It looks like we've already come to this conclusion for request telemetry in the past. Seems like it never made its way to dependencies.

See:
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/33696c0d09a3c6e93a690198a09bd49d84a63c58/Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs#L168

And from https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/113cf9ec4aa3dfd0d1c6443ded1700a0fa65972b/CHANGELOG.md
>Response code 401 is part of the normal authentication handshake and will result in a succesfull request.

**Comment by zakimaksyutov** [comment]
I think requests and dependencies might have different logic in regards to 401. For requests 401 might be considered success. But for dependencies the authN/Z are not always done by impersonating users. And 4xx for dependency often leads to 5xx for requests. 

Common example - 3rd party service changed something and started to respond with 401/400/404. This is unexpected and as a result an app returns 500. In this case we would like to see rise in dependency failures.

**Comment by ShaneCourtrille** [comment]
@OsvaldoRosado Unfortunately I've come to realize that it's a bit more complicated.  A 401 is fine IF it's followed up by a 200 for the same request.  If not then a 401 is an error.  I'm curious if the AppInsights architecture would actually allow for this distinction though.

**Comment by tannguyen3** [comment]
I have a WPF app (.NET 4.6.1) that calls a Restful API and the HTTP Dependency collector only collects the 401 code of auth-challenge request and immediately reports as failed call. 
So I have 100% failed dependency but in reality everything works well. 

One thing I notice that, if Fiddler is opened, the collector is able to ignore 401 and collect the follow-up code after the auth-challenge request.
